### ISSUE TITLE
Incompatible with Keras >= 2.3 and tf.keras

### ISSUE THREAD

**Issue by a96tudor** [danielegrattarola/spektral#12]
Hi,

It seems that this is now incompatible with the latest Keras release (2.2.5) from August 22, 2019. Here are the details:

## Environment
Python 3.7
spektral==0.0.12
tensorflow==1.14.0
keras==2.2.5

## How to reproduce
Run a python script that imports `spektral`:
```python
import spektral
```

## Expected behaviour
Everything runs smoothly and `spektral` is imported correctly. 

## Observed behaviour
We get an import error:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<path_to_virtualenv>/lib/python3.7/site-packages/spektral/__init__.py", line 5, in <module>
    from . import layers
  File "<path_to_virtualenv>/lib/python3.7/site-packages/spektral/layers/__init__.py", line 4, in <module>
    from .convolutional import *
  File "<path_to_virtualenv>/lib/python3.7/site-packages/spektral/layers/convolutional.py", line 5, in <module>
    from keras.backend import tf
ImportError: cannot import name 'tf' from 'keras.backend' (<path_to_virtualenv>/lib/python3.7/site-packages/keras/backend/__init__.py)
```

## Suggested change
On the short run, I think it would suffice to pin the keras version in `setup.py` to 2.2.4 (This fixed the issue in my project for now)

Thanks,
Tudor.

**Comment by danielegrattarola** [comment]
Hi Tudor,

yes, this is probably because the original multi-backend Keras is undergoing some changes and will be superseded by tf.keras soon.
Rather than pinning the requirements to use an outdated version, I'll probably migrate the whole library to use tf.keras, since some parts of the code are already tensorflow-based. 

Unfortunately right now I am really busy and don't have much time to do the migration, but it will happen sometime in the coming weeks.
This will also coincide with a major version release of Spektral, since it's been stable for a while now.

So in short: thanks for opening this issue and stay tuned for coming changes in Spektral :)

**Comment by danielegrattarola** [comment]
By the way, this is fixed in commit 37fa74e on the `develop` branch.
Will merge into master soon, as well as publishing new PyPi version

**Comment by a96tudor** [comment]
Hi, @danielegrattarola 

Thanks a ton for the fix! I'm also happy to help out with transitioning towards `tf.keras` if you need that. 

Tudor.

**Comment by danielegrattarola** [comment]
Commenting on this issue because it seems appropriate.
Currently both Keras 2.3 and `tf.keras` (>=1.14.0) have some issues when dealing with sparse inputs, making them not suitable for graph neural networks.
See:
- https://github.com/tensorflow/tensorflow/issues/31819
- https://github.com/tensorflow/tensorflow/issues/32737

I have frozen the required version of Keras to 2.2.5, and am currently waiting to decide whether it's worth the hassle to move to TensorFlow 2 directly (which will probably have the same issue, plus others that make it difficult to develop like https://github.com/tensorflow/tensorflow/issues/31973).

Any suggestions are appreciated.

**Comment by cgarciae** [comment]
I think its worth the hassle because TF 2.0 + `tf.keras` are becoming the standard, it wouldn't be nice to split users into Keras users and `tf.keras` users. :)

**Comment by danielegrattarola** [comment]
@cgarciae I have tried to port the framework to TF2 several times, and I've always found that there is no clean way of addressing the lack of support for SparseTensors inputs. 

I will try another time because maybe things have been fixed, but currently, our best hope is to pressure the tf devs into restoring the functionality. 
Try opening issues or commenting on the issues linked above.

**Comment by cgarciae** [comment]
Seems good! Just saying it should be in scope :+1:

**Comment by danielegrattarola** [comment]
For those who were following this issue: I've found a way to run the GCN example in TF2 and Keras 2.3.1. 
I will post it soon, but the main difference consists of manually transforming the sparse adjacency matrix in a SparseTensor, and using the `steps` keyword instead of `batch_size` when calling `model.fit()`.
I'll be looking into the other settings as well, stay tuned.

`tf.keras` is a whole different story because it is not compatible with Keras, so I will need to decide whether to keep using Keras or move to `tf.keras`. If anyone has an idea on how to support both, e.g., via dynamic imports, let me know.

**Comment by cgarciae** [comment]
I see that the problem is in the functional API where

```python
x = tf.keras.Input(shape=(32,), sparse=True) 
y = tf.keras.layers.Dense(1, activation='sigmoid')(x) 
model = tf.keras.Model(x, y)     
```
doesn't work. But if you use a custom model (in this case `Dense` for simplicity)
```python
x = tf.sparse.SparseTensor([[0, 0], [0, 2]], values=[1, 1], dense_shape=[1, 3]) 
y = tf.keras.layers.Dense(5)(x)
```
it works as expected. Could a workaround be just not using the functional API while this is being fixed?

**Comment by danielegrattarola** [comment]
There currently is a [development branch for TF2 and tf.keras](https://github.com/danielegrattarola/spektral/tree/tf2) with a working implementation that should cover the majority of use cases. 

For all discussions and updates about the migration to TF2 see #10.
